### ISSUE TITLE
DWA compression slow for tiles larger than 64

### ISSUE THREAD

**Issue by axxel** [AcademySoftwareFoundation/openexr#344]
I noticed something strange (and showstoppingly critical for my application) with the run time behavior of compressing a tiled OpenEXR with DWAB compression: If I use exrmaketiled to recode a 100MP scanline file into a tiled file, the run time increases significantly if I increase the tile size. Here are a few measurements:

tile size  -> run time
64x64 -> 14s
128x128 -> 30s
256x256 -> 42s

This seems completely counter intuitive, since the smaller the tiles are, the more restarts and tile header generations, etc. are necessary. DWAA behaves the same. As a reference: when doing the same sequence with PIZ, I get the following timings: 8s, 4s, 3s. Which is what I would expect.

Before I start digging in an unknown code base, does anyone have an idea why the library behaves like that and whether or not this can be fixed (easily)?

**Comment by axxel** [comment]
@karlrasche may I bring this issue to your attention. I understand that you are the author of the DWA compression code, so maybe you have an idea what is going on here or where I might start looking? I'd appreciate any help you might offer. Thanks.

**Comment by cary-ilm** [comment]
Looking into the OpenEXR backlog. This warrants further investigation, so leaving the issue open for now.

**Comment by karlrasche** [comment]
Unfortunately, this sounds like it's working as expected. 

In tiled mode, DWAA/B (they're the same on tiles) switches from using the lossless Huffman coder from PIZ to zlib (like ZIP). This is more space efficient for smaller tiles, as the overhead from storing the symbol frequencies can swamp the actual pixel data. But, it's also considerable slower to decode. 

I'm fairly sure if you ran this through a profiler, we'll see most of the time spent in zlib.

**Comment by axxel** [comment]
Thank you @karlrasche for getting back to me on this one. But after reading your comment, I believe there is a misunderstanding. My 'complaint' is not about tiled vs non-tiled DWA compression but about small tiles vs. big tiles. Why should it take about twice as long to compress N tiles of size 128x128 compared to 4xN tiles of size 64x64? Note: the total amount of pixels to process is the same. Also I'm not talking about the decoding but the encoding part.

**Comment by peterhillman** [comment]
I did a test and found a similar pattern: increasing the tile size does increase compression run time, but also the file size drops 
![DWAB compression](https://user-images.githubusercontent.com/1069876/61019837-b3c42c00-a3ef-11e9-93af-6145c12ea4af.png)
16x16 tiles took 1.8s and produced a 7.1MB file, 128x128 took 7.9s but produced a 4.3MB file.
So, that extra time spent processing does result in a smaller file.
128x128 PIZ tiles compress significantly faster, but the file is larger, at 5.0MB.

Profiling shows it spent 6s in zlib for 128x128 tiles of the total 7s run time as @karlrasche predicts, but only 1s in zlib for 32x32 tiles.  It would appear that, even though the total number of pixels is the same regardless of tile size, zlib has more data to work with each time it compresses a large tile compared to a smaller one, so it takes proportionally longer to compress and can do a better job of compression as a result.

Since the time is spent in zlib, not in the DWA compression functions themselves, it's hard to imagine it would be possible to make significant improvements to compression speed without making the output files larger and also breaking compatibility. 

The advice might be to avoid DWAA/B compression for large tiles whenever compression speed is critical, but I don't think there's anything wrong with the compression code.

Note also that exrmaketiled is single threaded. In your own app you can get better performance by setting an appropriate globalThreadCount, and calling `writeTiles` rather than `writeTile` to compress many tiles simultaneously on different threads.

**Comment by axxel** [comment]
Thank you @peterhillman for your time. Based on your and @karlrasche's feedback I started looking into the code and found something very interesting: a one-line and backwards compatible patch completely fixes my timing issue _while_ reducing the file size considerably at the same time. A 4MP sample stored in 256x256 DWA compressed tiles gets a 17% size reduction compared to the current code while being compressed about 10 times as fast as before!

First I started playing with the compression 'strength' parameter of the `compress2` call, which is set to 9. I made the following measurements:
![dwa-time-vs-size-over-zipcompression](https://user-images.githubusercontent.com/1038346/61046844-69867f00-a3de-11e9-8bb6-4987284b194e.png)
As you can see, the run time (in s) increases steeply at around 5 or 6 while the file size (in MB) decreases only marginally. Note, the default setting for this parameter of the command line utility `gzip` is 6.

Based on that finding, I would have proposed to set that to 6 and achieve a speedup of 4x for getting a 3% larger file. Since we are only talking about tiled images and they are always larger that scanline images anyway and their whole purpose is faster access, I'd consider that a very reasonable trade off.

Then I had another thought: piz compressed scanline files are both a lot smaller and can be compressed faster than zip compressed files (which makes zip as a compression option completely pointless anyway?). This suggests that even though @karlrasche determined it is better to use zip instead of piz internally for the tiled case (because of the overhead of the piz header compared to the zip one) there must be a tile size where the balance tips over in favor of piz. It turns out this tipping point is 64x64 already:
![dwa-fix-timings](https://user-images.githubusercontent.com/1038346/61047952-e87cb700-a3e0-11e9-8b37-8862681062b5.png)
The `dwa*` simply initializes the `DwaCompressor` object in `ImfCompressor.cpp:215` with `STATIC_HUFFMAN` instead of `DEFLATE`. **So even the default size of 64x64 becomes twice as fast and 12% smaller.** And since the used compression scheme during decompression is determined based on the file contents and not based on whether it is tiled or not, this is backward compatible.

In summary, I'd like to propose:
 1. definitively select `STATIC_HUFFMAN` based on the tile-size (starting with 64x64)
 1. and optionally use zlib compression parameter 6 instead of 9 (also in the 'unknown data' part)

If this sounds good to you and you'd like me to prepare a PR, let me know.

**Comment by karlrasche** [comment]
Thanks Peter, that confirms what I suspected.

As memory serves, switching between STATIC_HUFFMAN and DEFLATE is recorded
in the bitstream, so switching the logic for larger tiles should’t break
compatability.

There’s also faster zlib replacements out there.  You could try dropping
one of those in and see how it performs.

Karl

**Comment by axxel** [comment]
@peterhillman: Since Karl's last comment ("Thanks Peter...") to my proposal for a fix suggests some confusion about who is talking to who, I'm not quite sure whether or not you consider his reply to be an answer to my (implicit) question ("interest in a PR from me?").

I'm not in a hurry but would certainly appreciate if this could get in before the next release. And thanks a lot to you and your colleagues who resurrected this 'dormant' project! :-)

**Comment by peterhillman** [comment]
Hi @axxel, Thanks for doing that investigation, and for the proposal. A PR would be good, particularly if you've already implemented your proposed changes.

I'm cautious about recommending adding this to OpenEXR in time for a next release, though.
Your proposal may well be a better choice than the original settings, but I'm nervous about a change in behavior between releases. Having a newer version of OpenEXR produce a larger file than an older one is unexpected behavior that may cause serious issues to people who are relying on small files.
Using zlib compression 6 instead of 9 will produce a larger file. In your test image STATIC_HUFFMAN produces smaller files than DEFLATE but that might not always be true. Image noise characteristics would be a significant factor in performance, so a wide variety of test images would be needed.

In any case, generally decompression performance is more important than compression performance, and we'd have to measure that too. We would also have to consider what (if any) changes it would be appropriate to make to scanline files.

Adding an extra header attribute to switch to the new scheme would be a workaround, so you can opt-in to the better performance, but seems a little messy. If there were addtional changes to be made to DWA compression which weren't backwards compatible, it would make sense to make DWAC/D compression types, which would use your proposed settings, and we could then treat DWAA/B as legacy.

**Comment by karlrasche** [comment]
Peter -

If you’re worried about preserving behavior, could we have a thresholded
STATIC_HUFFMAN/DEFLATE switch on DWAB, and leave DWAA alone?

ATM for tiled files they behave the same. That would avoid introducing new
enums and should be backward/forward compatible.

Picking the right threshold is another issue though. I agree we need a
sampling of image content tested for encode/decode perf vs size to find the
right trade-off.

Karl

**Comment by peterhillman** [comment]
That had also occurred to me: It would still be an under-the-hood change, but at least there would be a way to get back the old performance by changing the compression type. However, there would be no way to change scanlines in a similar way, which seems worthwhile too.

Perhaps it would be sufficient to make a `#define` control, so that the new behavior is the default, but a single line change in a `.cpp` file could be made to restore the old performance if required.

Karl, do you have a feel for what the threshold should be for STATIC_HUFFMAN/DEFLATE switching for scanlines? Or should it always be STATIC_HUFFMAN for scanlines in both DWAA and DWAB?

**Comment by karlrasche** [comment]
DWAB with scanlines uses large enough strips that DEFLATE isn’t an issue,
so STATIC_HUFFMAN should be fine for that all the time.

DWAA with scanlines uses strip sizes that are ~PIZ (I cant remember which
uses 16 and which uses 32 off the top of my head between DWAA, PIZ, and
ZIP), so I don’t think using DEFLATE is an issue there either.

So I don’t think there’s a use for DEFLATE on scanlines - only for small
tiles.

**Comment by axxel** [comment]
I admittedly did my tests only on one 'random' image from my own camera/software setup. @karlrasche do you remember which tile sizes you tested and which image content you used when you decided to go with DEFLATE instead of STATIC_HUFFMAN originally?

I did not test the decompression speed but I expect it to be higher as well and even if it was not, I recon you would not want to change the scanline path to use DEFLATE to save a few percent on decompression if the compression part becomes 10x slower?

From my point of view, the current behavior is a 'bug': the inconsistency in both performance and file sizes between scanlines and tiles is undocumented, surprising and (as it turns out) an unnecessary pessimization. So I (personally) see no point in preserving that behavior.

If you'd want to make sure people could still opt into the DEFLATE path (for all tile sizes) by making a distinction between DWAA/B, that would certainly work for me. Although I would assume at some point in time someone might stumble upon that inconsistency and file another bug report.

I (personally) don't really care about the change of the DEFLATE 'strength' parameter for several reasons:
 * both runtime and size differences vanish the smaller the data set (tile size) gets and 32x32 is really small.
 * I don't care about tile sizes below 256x256 at all (and wonder which application would choose something like 32 in the first place).
 * I don't have any non RGB content where I would benefit from a 4x speed improvement in the 'unknown data' path.
That said, I thought it might benefit others.

I'll post a PR without the 9->6 change and hope I can make a convincing case for applying it as a 'bug fix'. The exact size threshold can be adjusted easily if 64 should not be best choice (as long as it is blow 256 I do (again) not really care).

**Comment by karlrasche** [comment]
As memory serves, I was doing large sweeps of  tile sizes on various types
of renders, looking an en/decode tiles and file sizes.

STATIC_HUFFMAN should beat DEFLATE in decode, the same way PIZ is faster
than ZIP. If anything, there may be more optimization possible on the
Huffman encode side - I only optimized the decoder.

Switching the coder is IMHO the right solution.

Smaller tiles are generally better for packing large texture footprints
into texture caches, but the “best” tile size for such a thing is
complicated. Not sure if anyone is rendering from DWAx tiled files, but it
might be good to check.

**Comment by axxel** [comment]
Is there any interest in dealing with this issue before the 3.0 release? (see also my last comment from September in #445 )

**Comment by peterhillman** [comment]
I think 3.0 is the right time to do this. A 3.1 is planned shortly after 3.0, so there's a timeline to address any issues that arise. As it happens, there aren't any other changes in the layout with 3.0, but making this change in a major release would make it less likely to be missed.

@karlrasche, one thought: `exrmaketiled` and OpenImageIO's `oiiotool` and `maketx` all default to tiles of 64x64. Your PR changes the encoder above 48x48, so this change would affect anyone writing DWA tiled files without changing the default size. This means that DWA tiled OpenEXR-3.0 files written with default options are no longer binary identical with older versions. As you say there is a performance gain from switching even at 64x64, but I wonder whether changing the threshold to above 64x64 might be a safer option?

**Comment by peterhillman** [comment]
Actually, I take that back: I was forgetting #846: before 3.0, a random unintialized byte was appended to each tile, which prevented files being binary identical. That means it isn't so important to maintain the behavior of default/64x64 tiles across versions.

@axxel, would you like to update your PR to resolve the conflicts? The source file structure has been reorganized since that PR.

(also, apologies: in my previous post I'd misread who had posted which comment to the thread)

**Comment by lgritz** [comment]
Are we talking about the *files* not being binary identical, or the *pixels*? (Most people care a lot more about the latter.)

A mitigating factor: While some studios might use lossy DWA compression in textures, I think it's not very common (my place has never done it and the OIIO testsuite and docs don't do it). So I think even if the 64x64 tiles change slightly, because it is only when using this compression mode, not many people are likely to be inconvenienced.

**Comment by peterhillman** [comment]
I think the decoded pixels stay the same. The changes here are to the lossless part of the compression, rather than the lossy. So, the decoded values should be exactly the same as before.

There are folks who compare checksums to confirm that nothing has changed between releases of tools which write OpenEXR files. Before #846, a test which writes DWA images may well have got inconsistent checksums due to the uninitialized byte, but #846 and this change will change both file size and checksums for tiled DWA images. Such a change might be anticipated across a major version bump, so if it's documented it shouldn't be a problem.

**Comment by axxel** [comment]
Thanks for looking into this again. I can certainly update the PR. Since the discussion about this topic has been a while ago and lengthy, I'd like to recap what I remember to be important:
* DWA(A/B) uses STATIC_HUFFMAN for scanlines but DEFLATE for tiles
* Both result in exactly the same pixel values after decompression.
* For (IMHO) irrelevantly small tile sizes of less than 64x64 pixels DEFLATE is better in terms of size and speed (see table above)
* For a specific image (sample from my camera based HDR content) with RGB half pixel format the crossover point was 48² pixels.

Based on that I see 3 valid alternatives to improve the situation:
 1) use the patch I proposed (where the switch point is fixed at a tile size in bytes which corresponds to 48x48 RGB half pixels)
 2) use DWAA for DEFLATE and DWAB for the STATIC_HUFFMAN (or vice a versa)
 3) switch to STATIC_HUFFMAN unconditional

Number 1) is user friendly and produces likely the best results but is a bit 'random' and does not produce the 'ideal' output for every image near the switch point. 2) seems very reasonable to me and would be similar in spirit to their meaning in the scanline case but is not really user-friendly especially when this choice ends up to be one for the 'end user'. 3) would obviously be the simplest and most likely give 'the best' performance in space and time for realistic tile sizes.

For my personal use case they are equivalent. I guess 1) or 3) are best for the 'average' user.

**Comment by axxel** [comment]
Anyone any preference as to which path my updated PR should take (1, 2 or 3)?

**Comment by lgritz** [comment]
If the pixel values stay the same, absolutely no preference from me.

**Comment by peterhillman** [comment]
I'm liking 2. It makes any changes opt-in or opt-out, which gives a simple path to anyone who wants to use DEFLATE for consistent behavior or because they've found a case where it performs better.
Compression choices are indeed a little esoteric and user-unfriendly, but changing the behavior of DWAA vs DWAB for tiles doesn't seem less user-friendly than the existing difference for scanline files. Software which writes OpenEXR files should ideally provide sensible defaults, or simplify choices into things like a 'use lossy compression' checkbox, and only offer fine-grained control over compression settings etc in an 'advanced options' mode.

**Comment by karlrasche** [comment]
2 seems reasonable to me as well (at least until we need to add another
dimension of compression options!)

Karl
